Toolbox Beta Complete

DanVanAtta opened the pull request:
## Overview
This is the moderator toolbox with tabs:
- moderators
- banned users
- banned user names
- api keys
- access log

Functionality added for moderators:
- can view access log
- can add user bans
- can add username bans
- can create single use-keys (to register on new machines)
- can delete api keys
- can view moderators

Functionality added for super-moderators:
- can promote users to moderator
- can promote moderators to super-mod
- can demote moderators to a user

### DB changes
- add "super_mod" column to lobby_users
- rename "mac_ban" table to "user_bans"
- remove moderator columns from "user_bans"
- add "last used host" to moderator IP key (to support audit)
- add host column to single use api key table (for audit)
- resurrected lobby access log table (reverted the removal)

### Non Toolbox Changes
- 'isBanned' method now returns just ban expiry instead of tuple. We can now remove bans directly and can simplify this.
- removed functionality to remove a ban by setting ban expiry in the past. Because we can remove bans, this clever trick is no longer needed.

## Other notes
- A test-only endpoint was added to clear lockouts, this is to support testing.
- ***Important*** A script was added to lobby-db to load sample data. This loads a super-mod user "test" with password "test" and is prepped with a single-use-api key with value "test".

## Screenshots
![Screenshot from 2019-06-16 13-17-52](https://user-images.githubusercontent.com/12397753/59569102-d013b980-9039-11e9-86a7-3907e5688722.png)
![Screenshot from 2019-06-16 13-17-59](https://user-images.githubusercontent.com/12397753/59569103-d013b980-9039-11e9-80e6-48c1a3368b25.png)
![Screenshot from 2019-06-16 13-18-07](https://user-images.githubusercontent.com/12397753/59569104-d013b980-9039-11e9-8492-c285263478ee.png)
![Screenshot from 2019-06-16 13-18-29](https://user-images.githubusercontent.com/12397753/59569106-d0ac5000-9039-11e9-9d14-a422dba107d7.png)
![Screenshot from 2019-06-16 13-18-55](https://user-images.githubusercontent.com/12397753/59569107-d0ac5000-9039-11e9-8e5a-c8d294cc82a6.png)
![Screenshot from 2019-06-16 13-19-07](https://user-images.githubusercontent.com/12397753/59569108-d0ac5000-9039-11e9-9ead-03d1eb2b8834.png)
![Screenshot from 2019-06-16 13-18-19](https://user-images.githubusercontent.com/12397753/59569105-d0ac5000-9039-11e9-9730-60464d37d78b.png)
![Screenshot from 2019-06-16 13-19-50](https://user-images.githubusercontent.com/12397753/59569109-d0ac5000-9039-11e9-826d-10c68f495b13.png)
![Screenshot from 2019-06-16 13-19-57](https://user-images.githubusercontent.com/12397753/59569110-d144e680-9039-11e9-97d1-b36d02795294.png)
![Screenshot from 2019-06-16 13-20-06](https://user-images.githubusercontent.com/12397753/59569111-d144e680-9039-11e9-9072-8faa5e0507c4.png)
![Screenshot from 2019-06-16 13-20-15](https://user-images.githubusercontent.com/12397753/59569112-d144e680-9039-11e9-8543-6c7289b419af.png)
![Screenshot from 2019-06-16 13-20-30](https://user-images.githubusercontent.com/12397753/59569114-d144e680-9039-11e9-992f-f4444039b2ba.png)

## Additional Review notes

There are some outstanding TODO items: https://github.com/triplea-game/triplea/issues/4894
I'll try to get back to them when I return from traveling.

DanVanAtta commented:
@RoiEXLab @ron-murhammer I'd like to get this merged Saturday or before so that it is put to bed before my travel dates where I'll be off-grid for close to two months. If some, perhaps even a quick review can't be done before then, I think that means skipping or abbreviating it.

RoiEXLab commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/AccessLogTab.java:
```diff
@@ -0,0 +1,75 @@
+
+    final JButton loadMoreButton = JButtonBuilder.builder()
+        .title("Load More")
+        // .actionListener(() -> accessLogTabActions.loadMore(table))
```
Not sure if this makes sense here, but we agreed that commented out code might as well be removed

RoiEXLab commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/AccessLogTabActions.java:
```diff
@@ -0,0 +1,179 @@
+        .filter(AbstractButton::isSelected)
+        .findFirst()
+        .map(AbstractButton::getText)
+        // .map(TIME_DURATION_MAP::get)
```
Same here with commented out code

RoiEXLab commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/BadWordsTabModel.java:
```diff
@@ -27,15 +31,30 @@ String removeBadWord(final String wordToRemove) {
+   * Returns a "2 x N " list of lists. The first column value is the bad word value from the bad words table,
+   * the second column value is the "remove" button.
+   */
+  List<List<String>> getTableData() {
```
I must admit that I'm not really convinced this is the best data structure for a table, but it works so whatever

(To clarify, the column amount is usually fixed, so a different inner type is usually the way to go, which also supports typisation etc.)

RoiEXLab commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/EventLogTabModel.java:
```diff
@@ -22,10 +22,10 @@
 class EventLogTabModel {
   @VisibleForTesting
-  static final int PAGE_SIZE = 10;
+  static final int PAGE_SIZE = 30;
```
👍 For pagination, even though I'd have hoped there was some API for this in the standard lib for the client side

DanVanAtta replied:
Interesting comment. The paging needs to make its way through to DB, just as the client does not want to get all table data, the server should not read all data either. Without a stateful server, I have trouble seeing how it could be standardized.

RoiEXLab replied:
I was more thinking about the UI support of such a concept.
Having coded a lot of react the last year, I noticed there are a lot of small things like support for this in specific libraries that make your life easier.

RoiEXLab commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/ModeratorsTabModel.java:
```diff
@@ -1,22 +1,69 @@
+                modInfo.getName(),
+            Optional.ofNullable(modInfo.getLastLogin())
+                  .map(Instant::toString)
+                .orElse("")))
```
Like mentioned earlier in the data structure comment: Having different types here would make this more readable.

DanVanAtta replied:
It would be cool to have a `JTable<ObjectType>`  where it is initialized to have a mapper that could convert an `ObjectType` to a table row. I think that is a nice-to-have, probably better to ensure that the needed controls are present and that there are no errors in the conversion, nor is there any missing data. I do like that idea/thinking. I wonder if it's worth it long-term and jives with a javafx API. My hope is that we can convert to swing builders and then change the underlying implementation of the builds to produce JavaFx components instead. That way we are not updating any code outside of swing-lib when migrating away from swing.

RoiEXLab commented on http-client-shared-data/src/main/java/org/triplea/http/client/JsonEncoder.java:
```diff
@@ -0,0 +1,27 @@
+public class JsonEncoder implements Encoder {
+  private static final GsonEncoder gsonEncoder = new GsonEncoder();
+
+  @SuppressWarnings("deprecation")
```
We already had this in the other PR, although I forgot the exact details, but wouldn't this already be a follow-up PR?
Fixing this should be rather straight forward, a simple wrapper object is required.

Regardless I wouldn't suppress the warning if the fix is so simple.

DanVanAtta replied:
My 2 cents, a follow-up PR would have that as a dedicated change so that we're not debugging multiple things at the same time. I'm also not sure if/when I'll get to this, using the deprecated API here does not bother me, I don't see the real benefit so long as we are on the current version of feign. In principle I do support fixing this and would/will if I can get to this level of priority. I'm pushing to get all the testing written.

RoiEXLab commented on http-client-shared-data/src/main/java/org/triplea/http/client/moderator/toolbox/ModeratorToolboxFeignClient.java:
```diff
@@ -0,0 +1,69 @@
+  @Headers({HttpConstants.CONTENT_TYPE_JSON, HttpConstants.ACCEPT_JSON})
+  RegisterApiKeyResult registerKey(RegisterApiKeyParam registerApiKeyParam);
+  // @Param(ModeratorToolboxClient.SINGLE_USE_KEY_PARAM) String singleUseKey,
+  // @Param(ModeratorToolboxClient.NEW_API_KEY_PASSWORD_PARAM) String newPassword);
```
Code comment

DanVanAtta replied:
Not terribly surprised, only the code outside of toolbox or in the DAO layer is really ready for this detailed of code improvement/cleanup. I'd urge you back up on the review magnifying class, try to focus on if we have all the right data, if these are the right updates and not that the code is perfect. With that said, I do appreciate the head start on the review.

RoiEXLab commented on http-client-shared-data/src/test/java/org/triplea/http/client/moderator/toolbox/RegisterApiKeyResultTest.java:
```diff
@@ -0,0 +1,29 @@
+package org.triplea.http.client.moderator.toolbox;
```
Stop Review 1

DanVanAtta commented:
@RoiEXLab this is still a WIP, the DAO code is the most polished and tested, the rest might change a bit as I fix checkstyles and add tests. Hence the sections of commented out code. I'm working through the service and controller layers currently.

DanVanAtta commented on http-server/src/main/java/org/triplea/server/moderator/toolbox/banned/users/PublicIdSupplier.java:
```diff
@@ -3,11 +3,20 @@
   public String get() {
-    return Instant.now().getEpochSecond() + String.valueOf(Math.random()).substring(0, 5);
+    return Instant.now().getEpochSecond()
+        + String.valueOf(Math.random()).substring(0, 5);
```
Good mention, I had forgotten about the UUID class.

One thing to consider is the key length, 15 vs 36. Perhaps it's not a big deal to have the longer key, can give UUID a try.

DanVanAtta commented on http-server/README.md:
```diff
@@ -97,3 +147,9 @@ Essentially fail-fast:
+## 404 error, but endpoint is registered!?
+
+Make sure in addition to the `@Path` annotation on the endpoint method,
+ensure the controller class has a `@Path("")` annotation on it.
```
Do you know if they are equivalent?

RoiEXLab replied:
I don't know for sure but given the typical nature of root paths for URLs (http://example.com vs http://example.com/ ) I'd assume it for this case.

DanVanAtta commented on http-server/src/main/java/org/triplea/server/http/IpAddressExtractor.java:
```diff
@@ -25,6 +25,6 @@ public static String extractClientIp(final HttpServletRequest request) {
       return forwarded;
     }
-    return request.getRemoteHost();
+    return request.getRemoteAddr();
```
Addr is IP address, host is the hostname. I was seeing null values from 'getRemoteHost', I think that is probably the reason why.

RoiEXLab replied:
Good to know, I always thought if there's no reverse DNS set, there would be some generic one from the ISP.

prastle commented:
Off topic but is anyone having issues logging into the forum or lobby? @DanVanAtta @RoiEXLab ?

DanVanAtta commented on http-server/src/main/java/org/triplea/server/moderator/toolbox/api/key/GenerateSingleUseKeyService.java:
```diff
@@ -0,0 +1,48 @@
+    final Integer userId = userLookupDao.lookupUserIdByName(moderatorName)
+        .orElseThrow(() -> {
+          log.log(Level.SEVERE, "Failed to find user by name: " + moderatorName);
+          return new IllegalStateException("No user found by name: " + moderatorName);
```
The uncaught exception would be logged on server side, I think you're correct it would be logged twice.

DanVanAtta commented on http-server/src/main/java/org/triplea/server/moderator/toolbox/api/key/registration/ApiKeyPasswordBlacklist.java:
```diff
@@ -0,0 +1,41 @@
+      "login",
+      "1111",
+      "11111",
+      "!@#$");
```
Set makes sense to use. The lookup time is nano second whether via hash or iteration.
c0cf6cbab

DanVanAtta commented on lobby-db-dao/src/main/java/org/triplea/lobby/server/db/TimestampMapper.java:
```diff
@@ -0,0 +1,30 @@
+  public static Instant map(final ResultSet resultSet, final String columnName) throws SQLException {
+    final Calendar cal = Calendar.getInstance();
+    cal.setTimeZone(TimeZone.getTimeZone("UTC"));
+    return
```
it is interesting formatting, fixed in: ae386f2a3

DanVanAtta commented on lobby-db-dao/src/main/java/org/triplea/lobby/server/db/dao/ModeratorAuditHistoryDao.java:
```diff
@@ -20,6 +27,31 @@
+
+    BAN_USER,
+
+    REMOVE_USER_BAN, REMOVE_MODERATOR, ADD_SUPER_MOD,
```
yeah, that's the auto-generated enum values being added vs hand-formatting. all the audit actions should be defined now, I can go ahead and fix this.

DanVanAtta replied:
11c4a7d96

DanVanAtta commented on lobby-db-dao/src/test/resources/datasets/user_lookup/select.yml:
```diff
@@ -1,10 +1,11 @@
-moderator_action_history:
 moderator_api_key:
+moderator_action_history:
+moderator_single_use_key:
```
I've yet to go through the checkstyles, agree otherwise to fix:
a802df097

DanVanAtta commented on lobby-db-dao/src/main/java/org/triplea/lobby/server/db/dao/ModeratorKeyRegistrationDao.java:
```diff
@@ -0,0 +1,63 @@
+  @Transaction
+  default void invalidateSingleUseKeyAndGenerateNew(Params params) {
+    params.checkState();
+    final String publicKeyId = Instant.now().getEpochSecond() + String.valueOf(Math.random()).substring(0, 6);
```
They are logically the same thing, it should be the same code. I moved the public id supplier class to the DB dao code: 4c1e5f2d07479a3b792278d6f5b98d4f17862e0f

DanVanAtta commented:
@prastle please create new issues for off-topic conversations.

DanVanAtta commented:
100% please create new issues @prastle rather than finding the latest thread. It's super annoying and distracts the original issue and creates spam, adds work to cut a new issue and then reference others. Just create a new one.

For lobby issues like this also do feel free to text me.

tvleavitt commented:
There are plenty of systems out there with no forward or reverse DNS
defined, not surprised that getRemoteHost would return null values
frequently. There will always, of course, be an IP address.

On Fri, Jun 14, 2019 at 4:10 PM Dan Van Atta <notifications@github.com>
wrote:

> *@DanVanAtta* commented on this pull request.
> ------------------------------
>
> In
> http-server/src/main/java/org/triplea/server/http/IpAddressExtractor.java
> <https://github.com/triplea-game/triplea/pull/4899#discussion_r294019022>:
>
> > @@ -25,6 +25,6 @@ public static String extractClientIp(final HttpServletRequest request) {
>      if (forwarded != null) {
>        return forwarded;
>      }
> -    return request.getRemoteHost();
> +    return request.getRemoteAddr();
>
> Addr is IP address, host is the hostname. I was seeing null values from
> 'getRemoteHost', I think that is probably the reason why.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/4899?email_source=notifications&email_token=ABB5HAJD7FPAG353AZAS6YLP2QQPRA5CNFSM4HXXMPQ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3USBFQ#discussion_r294019022>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABB5HAPCHKJYM3RHN2NBAVDP2QQPRANCNFSM4HXXMPQQ>
> .
>


-- 
Thomas Leavitt
Internet enabled since 1990

tvleavitt commented:
What's the point of doing this at all, much less hard coding into source?
Any set here is going to be entirely random in actual practice. If you're
going to do this, have it be defined externally in a configuration file,
and then build it from a logical source, such as perhaps, the 100 most
frequent passwords seen on HaveIBeenPwned or something... I don't see the
value of blocking "password1" but not "password123", "abcd" or "qwerty" but
not "asdf", etc.

On Fri, Jun 14, 2019 at 4:16 PM Dan Van Atta <notifications@github.com>
wrote:

> *@DanVanAtta* commented on this pull request.
> ------------------------------
>
> In
> http-server/src/main/java/org/triplea/server/moderator/toolbox/api/key/registration/ApiKeyPasswordBlacklist.java
> <https://github.com/triplea-game/triplea/pull/4899#discussion_r294019876>:
>
> > +      "api-key",
> +      "1234",
> +      "pass",
> +      "password",
> +      "password1",
> +      "passw0rd",
> +      "qwerty",
> +      "abcd",
> +      "monkey",
> +      "batman",
> +      "football",
> +      "baseball",
> +      "login",
> +      "1111",
> +      "11111",
> +      "!@#$");
>
> Set makes sense to use. The lookup time is nano second whether via hash or
> iteration.
> c0cf6cb
> <https://github.com/triplea-game/triplea/commit/c0cf6cbab08c96d6bbae67a5f7c1aa59ef20d0a2>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/4899?email_source=notifications&email_token=ABB5HAOO2W4TYAFHU6LYAE3P2QRF5A5CNFSM4HXXMPQ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3USJFQ#discussion_r294019876>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABB5HAOKMYOGHE6S654P3XDP2QRF5ANCNFSM4HXXMPQQ>
> .
>


-- 
Thomas Leavitt
Internet enabled since 1990

DanVanAtta commented:
@tvleavitt I was debating a bit if there was enough value for an api-key password blacklist. I was a bit concerned that a simple password might be chosen. If that happens, and an api-key is compromised (on windows it's stored in registry, unix and mac it's on the OS somewhere but I'm not sure where), then an easy key would be full compromise. Given lockout on failed attempts and the slow hashing algorithm, it should make it really hard to brute-force a password+api-key combination, even if someone knows the api-key, or even worse knows both the api-key and hashed value in database

The password list is actually seeded from a list of the common top N passwords. Hence it's not super exhaustive, it's largely to keep mods from using "1234" and "password" as their password.

To be sure the list is to for the moderator api-key only, does not affect the lobby password.

As config vs hard-coded, we'd likely do a redeployment to update the config list and would otherwise need a restart. We could put the list in DB, but then it's great to have DB updates be done via checked-in source. Hence, an external config list or DB table doesn't seem necessary, the secrecy is a bit immaterial.

tvleavitt commented:
Dan,

I have a simple solution that will never need updating and will prevent any
and all stupid passwords: require the password to be 12 characters. That's
it.

I've played with this on HaveIBeenPwned's dataset... not only
mathematically is a longer password more secure, but humans are freaking
lazy as hell, so that a longer password is simply less likely to have been
previously used. If 1 million people have used password1, 100,000 used
password1, 1,000 used password123, and 100 used password1234. That held
through for multiple variations when I tested them. In actual practice,
simply adding 1234 to the end of any password makes it almost certain that
it won't be in that database. The number of people using abcd vs abcde or
abcdef falls off at a high rate as well.

Do that for the lobby password as well, and without any annoying complexity
requirements in terms of numbers, caps vs no caps, or special characters,
you've pretty much guaranteed everyone has a secure password.

Regards,
Thomas Leavitt



On Fri, Jun 14, 2019 at 10:03 PM Dan Van Atta <notifications@github.com>
wrote:

> @tvleavitt <https://github.com/tvleavitt> I was debating a bit if there
> was enough value for an api-key password blacklist. I was a bit concerned
> that a simple password might be chosen. If that happens, and an api-key is
> compromised (on windows it's stored in registry, unix and mac it's on the
> OS somewhere but I'm not sure where), then an easy key would be full
> compromise. Given lockout on failed attempts and the slow hashing
> algorithm, it should make it really hard to brute-force a password+api-key
> combination, even if someone knows the api-key, or even worse knows both
> the api-key and hashed value in database
>
> The password list is actually seeded from a list of the common top N
> passwords. Hence it's not super exhaustive, it's largely to keep mods from
> using "1234" and "password" as their password.
>
> To be sure the list is to for the moderator api-key only, does not affect
> the lobby password.
>
> As config vs hard-coded, we'd likely do a redeployment to update the
> config list and would otherwise need a restart. We could put the list in
> DB, but then it's great to have DB updates be done via checked-in source.
> Hence, an external config list or DB table doesn't seem necessary, the
> secrecy is a bit immaterial.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/triplea-game/triplea/pull/4899?email_source=notifications&email_token=ABB5HAKH6CFUTBVCKOSU2K3P2RZYHA5CNFSM4HXXMPQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXYQNOA#issuecomment-502335160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABB5HAMTWGIHLBFVGDLEDB3P2RZYHANCNFSM4HXXMPQQ>
> .
>


-- 
Thomas Leavitt
Internet enabled since 1990

RoiEXLab commented:
@DanVanAtta There are libraries like [zxcvbn](https://github.com/dropbox/zxcvbn) that can help estimate password strength. I have used this library in the past for a Web project of mine and was pretty satisfied. Even though this library is written for javascript, it has links to several ports to different languages there.

DanVanAtta commented:
My thinking was we want to defend against trivial passwords, but allow ones that are super easy to remember. Hence even mentioning that the lobby password could be shared, to some extent making it easy to remember is more important, I was tempted to not even require a password at all and just use the API key alone. We really do not want moderators to forget their password. The API key salted with password will be hard to crack.

DanVanAtta commented:
@RoiEXLab I've finished up most of the test code and it drove some of the changes that were also updated. I really leaned on experience to know that the code I was writing would have been testable, when in a rush I'll write more production code and interleave test a bit later rather than a more heavy TDD approach I would tend to favor otherwise.

Left to do:
- http client integration testing, wire mock test and client testing against a live server. 
- Add moderator feature is not wired fully through. I added the backend component for it, I think to add it in will be not very much code at all.

I'm hoping I'll be able to knock those elements out then focus on fixing build issues and self-reviewing and refining the final code.

Please be sure if you haven't checked already, the updates to the existing non-toolbox code that was updated. The biggest risk for problems I think is in those areas, but overall the updates there I think were relatively straight forward and strategic (if not just mixed in with a ton of new code).

Thanks again for working with me on this update and it landing as a very large 3rd iteration.

DanVanAtta commented:
@RoiEXLab thanks for the approve, I think this could be merged tomorrow :shipit: 

There are some notable updates that are perhaps worth taking a looking at. I added more testing, the tests are done! I broke up the client side http client, as one class it was overwhelming to test it. I also found a nice way to break up the controller integration tests by adding a custom Extension class that starts the server up just once. From there it was pretty fast to write integration tests for all of the controllers.

Remaining, before merge, are pretty small items that I hope to knock out in an hour or two after getting up tomorrow:
- checkstyles, number of javadocs to add
- any other build issues
- i'm seeing a single db-dao test failure, need to fix that
- finish wiring the add moderator feature and manual testing. The rest of the toolbox has been manual tested and after the completion of tests, it seems to all be working pretty okay.

There are some usability issues that remain but I think they'll have to be tackled post-merge and perhaps when I get back from travel:
- the issue noted in the TODO ticket.
- there are some usability issues that could use improvement. Like seeing which users are super-mod and disabling the remove mod and add super mod button for them (those actions always fail, super-mods can't be promoted or demoted).
- it would also be good to disable the ban name or ban user button if a user is already banned

Tomorrow I'll add usability issues to the TODO ticket after a final round of testing and depending on what I fix on the spot.

DanVanAtta commented:
Might be worth mentioning, in case testing locally, the DB data sample in docs/dev/README.md is updated so that the single-use api key value is now "test"

DanVanAtta commented:
@tvleavitt Certainly agree to not require special characters, too long of a password I think could run in conflict with making it easy to remember. For the case of lobby logins, there is no reset password feature, it's been important for users to remember their passwords : )

The moderator password is kinda valuable, when it comes to API key we pretty much have a very long password as the key and password are concatenated together and then hashed. The password then serves the purpose of providing some defense against cracking in case the key and/or hash are compromised, but most importantly it provides a layer of validation in case someone forgets to log out of a system that they are letting someone else play on. Notably for example a moderator lets their family member play, and perhaps that family member is a very young kid who then goes to town executing moderator actions. Unlikely, but may as well not have the temptation.

I think the goal for at least api key password is mainly that, to actually try and crack the password is tough provided the password is not trivial.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/CreateNewApiKeyActions.java:
```diff
@@ -16,23 +18,20 @@
+    final RegisterApiKeyResult result = registerNewKeyClient.registerNewKey(apiKeyPassword);
 
-    return result.getNewApiKey()
+    return Optional.ofNullable(result.getNewApiKey())
```
@RoiEXLab API was updated to return non-optional values, that was important for JSON encoding to not encode optional objects into the JSON

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/Pager.java:
```diff
@@ -0,0 +1,52 @@
+ * requesting data will increment the current row to request the next row
+ * on the next data fetch.
+ */
+public class Pager {
```
This class was to avoid re-writing the same test code, @RoiEXLab, I think this is what you were looking for in terms of client-side support  : )

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/access/log/AccessLogTabModel.java:
```diff
@@ -0,0 +1,61 @@
+  private final ToolboxUsernameBanClient toolboxUsernameBanClient;
+
+  static List<String> fetchTableHeaders() {
+    return Arrays.asList(
```
Column header definitions were moved to model. This made it easier to line up the column values with the headers.

DanVanAtta commented on game-core/src/main/java/games/strategy/engine/lobby/moderator/toolbox/tabs/moderators/ModeratorsTab.java:
```diff
@@ -0,0 +1,115 @@
+        .toolTip("Promotes a user to the moderators")
+        .build();
+
+    // This is a very fancy listener that will validate the entered requested new moderator
```
@RoiEXLab not sure if you saw this bit of code, it's kinda interesting as the 'add moderator' button does not enable until the server has verified the user exists (this is done in background while user is typing). When testing this out I noticed the background thread was needed to not interrupt typing.

DanVanAtta commented on http-clients/src/main/java/org/triplea/http/client/moderator/toolbox/api/key/ToolboxApiKeyClient.java:
```diff
@@ -0,0 +1,78 @@
+  /**
+   * To only be used in non-prod (support testing).
+   */
+  public void clearLockouts() {
```
This is a very fishy thing to have... The clear lockout is only allowed in non-prod, but does not even require a valid API key. That means anyone can clear lockouts on our pre-prod server. I think it's probably okay as the data there is probably going to be well known / documented, and I suspect will not be abused in actuality. If so, we can add more security to enable this feature only in test (which would mean a third configuration file potentially)

DanVanAtta commented on http-clients/src/main/java/org/triplea/http/client/moderator/toolbox/register/key/RegisterApiKeyResult.java:
```diff
@@ -41,12 +44,4 @@ public static RegisterApiKeyResult newErrorResult(final String errorMessage) {
         .build();
   }
-
-  public Optional<String> getNewApiKey() {
```
As noted, the optional getters messed with JSON encoding.

DanVanAtta commented on http-server/src/main/java/org/triplea/server/moderator/toolbox/api/key/InvalidKeyLockOut.java:
```diff
@@ -44,14 +48,29 @@ public InvalidKeyLockOut(
+   */
+  public synchronized void clearLockouts(final HttpServletRequest request) {
+    if (production) {
+      // if accessed in production, not allowed, immediately lock out.
```
Bit noteworthy, if clearLockouts is accessed in prod, we do an immediate lockout of that IP.

DanVanAtta commented:
@RoiEXLab I've removed the WIP label. If you get a chance, a final pass could be useful, otherwise I'll plan to merge this tomorrow. The timing is good as I think I'll be lucky to have more than an hour or two more to dedicate to this effort before heading out for travels.

@prastle @ron-murhammer I've added screenshots, please review the overview to familiarize yourselves with the new feature if you would. The toolbox is still behind the "show test features" flag in settings. Likely that will be removed by the next release as the http-server will be production. I don't think there is too much left to do there other than to deploy certs and make the server be https.

RoiEXLab commented:
@DanVanAtta So there's the problem that since you sqashed all the commits I don't have a changeset of your recent commits, so I'd have to go over 9000 lines of code.
I don't think I'll be able to do that until tomorrow, I'll try to have a quick look, but the scrolling alone will take ages already

DanVanAtta commented:
@RoiEXLab I wont' have time to deal with any post-merge issues, I'm going to go ahead and merge. I'd rather leave this open for a day, but am kinda concerned that we might be reviewing for the wrong things too. At the risk of sounding ungrateful, I want to emphasize we should be reviewing for the right code, not the best code.

The pre-squash updates were to javadoc, the wiring of add-moderator was actually already half done at the service+DAO level, so the latest updates were really not terribly significant. I'm most concerned about the changes to the existing code. I also think the bar is different for maintainers who are on the hook for fixing items before releases and those that can more or less walk away. With that said, we need to be sure we're merging things with high enough confidence, quality, and testing so that we don't have a huge volume of things to fix at a much later date.

We also need to I think address eventually that review is new to TripleA. For most of TripleA's existence it's been one or two people jamming out code with no review. If we were to review that code today TripleA would have taken far longer to develop (but perhaps avoided the low quality, no testing quagmire we find ourselves in and are slowly digging our way out of). I'm not sure how to best balance it, with only a few developers, all non-professional, it's not too hard to outpace review and that creates a bottleneck. I think my point is we need to emphasize the right degree of efficiency.

With all that said, please do feel free to continue commenting on this diff @RoiEXLab . I can add to the TODO list to go over and address any additional items. Beyond today, I don't think I'll have any time until I get back and am focusing on travel prep.
